RaftMongoWithRaftReconfig.tla

```diff
@@ -0,0 +1,276 @@
+--------------------------------- MODULE RaftMongoWithRaftReconfig --------------------------------
+\* This is the formal specification for the Raft consensus algorithm in MongoDB.
+\* It allows reconfig using the protocol described in Raft.
+
+EXTENDS Naturals, FiniteSets, Sequences, TLC
+
+\* The set of server IDs
+CONSTANTS Server
+
+\* Server states.
+\* Candidate is not used, but this is fine.
+CONSTANTS Follower, Candidate, Leader
+
+\* A reserved value.
+CONSTANTS Nil
+
+----
+\* Global variables
+
+\* Servers in a given config version.
+\* e.g. << {S1, S2}, {S1, S2, S3} >>
+VARIABLE configs
+
+----
+\* The following variables are all per server (functions with domain Server).
+
+\* The server's term number.
+VARIABLE currentTerm
+
+\* The commit point learned by each server.
+VARIABLE commitPoint
```

Comment:
Can we model this as a global commit point? Advancing the global commit point is equivalent to promising a client the entry is committed. Once it's observed by a client (a global property), the entry cannot be rolled back on any node. Thus we don't have to model commit point propagation.

Modeling the committed entries as a set like Will did is another idea, which may make checking the no-op gets committed before reconfig easier.

Reply:
I tried to model this as a global commit point, but since Reconfig checks whether the previous config entry is committed, if the global commit point is in a higher term, we could incorrectly think the config entry is committed. I agree that modeling the commit point propagation feels more complicated than necessary for this spec. I like the idea of modeling the committed entries as a set, and I'll give it a try. Though I might prefer modeling commit point per node due to its similarity to your existing spec and similarity to the implementation.

I'll post another patch trying out modeling the committed entries as a set.

Reply:
I see. Good point.